core/include/scipp/core/variable_keyword_arg_constructor.h

```diff
@@ -0,0 +1,196 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2019 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Igor Gudich
+
+#ifndef SCIPP_VARIABLE_KEYWORD_ARG_CONSTRUCTOR_H
+#define SCIPP_VARIABLE_KEYWORD_ARG_CONSTRUCTOR_H
+
+#include "scipp/core/except.h"
+#include "scipp/core/vector.h"
+#include <type_traits>
+
+namespace scipp::core {
+
+// The structs needed for universal variable constructor are introduced below.
+// Tags are used to match the corresponding arguments treating the arbitrary
+// order of arguments in the constructor, and not mixing values and variances.
+// Structures Values and Variances just forwards the arguments for constructing
+// internal variable structure - array storage.
+
+namespace detail {
+
+template <class U> struct vector_like {
+  std::vector<U> data;
+  template <class... Args>
+  vector_like(Args &&... args) : data(std::forward<Args>(args)...) {}
+  template <class T>
+  vector_like(std::initializer_list<T> init) : data(init.begin(), init.end()) {}
+};
+
+template <class Tag, class... Ts> struct TaggedTuple {
+  using tag_type = Tag;
+  using tuple_type = std::tuple<Ts...>;
+  tag_type tag;
+  tuple_type tuple;
+};
+
+struct ValuesTag {};
+
+struct VariancesTag {};
+
+template <class... Ts> auto makeArgsTuple(Ts &&... ts) {
+  return std::make_tuple<std::remove_reference_t<Ts>...>(
+      std::forward<Ts>(ts)...);
+}
+
+template <class T> auto makeArgsTuple(std::initializer_list<T> init) {
+  using iter = typename std::initializer_list<T>::iterator;
+  return std::make_tuple<iter, iter>(init.begin(), init.end());
+}
+
+template <class... Args>
+using MakeArgsTupleReturnType =
+    decltype(makeArgsTuple(std::declval<Args>()...));
+} // namespace detail
+
+using Shape = detail::vector_like<scipp::index>;
+using Dims = detail::vector_like<Dim>;
```

Comment:
I think this is only working by accident because `Dim` is not the same precision as `index`. Maybe use a tag such as an int as I suggsted?

Reply:
As you pointed out `Dim` is an enum, so this is not an issue.